pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,299 @@
+package ftp
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"gofr.dev/pkg/gofr/container"
+)
+
+var (
+	// errNotPointer is returned when Read method is called with a non-pointer argument.
+	errNotPointer = errors.New("input should be a pointer to a string")
+	ErrOutOfRange = errors.New("out of range")
+)
+
+// ftpFile represents a file on an FTP server.
+type file struct {
+	response ftpResponse
+	path     string
+	conn     ServerConn
+	name     string
+	offset   int64
+	logger   Logger
+	metrics  Metrics
+}
+
+// textReader implements RowReader for reading text files.
+type textReader struct {
+	scanner *bufio.Scanner
+	logger  Logger
+}
+
+// jsonReader implements RowReader for reading JSON files.
+type jsonReader struct {
+	decoder *json.Decoder
+	token   json.Token
+}
+
+// ReadAll reads either JSON or text files based on file extension and returns a corresponding RowReader.
+func (f *file) ReadAll() (container.RowReader, error) {
+	defer f.postProcess(&FileLog{Operation: "ReadAll", Location: f.path}, time.Now())
+
+	if strings.HasSuffix(f.Name(), ".json") {
+		return f.createJSONReader()
+	}
+
+	return f.createTextCSVReader()
+}
+
+// createJSONReader creates a JSON reader for JSON files.
+func (f *file) createJSONReader() (container.RowReader, error) {
+	defer f.postProcess(&FileLog{Operation: "JSON Reader", Location: f.path}, time.Now())
+
+	res, err := f.conn.Retr(f.path)
+	if err != nil {
+		f.logger.Errorf("ReadAll failed. Unable to retrieve json file: %v", err)
+		return nil, err
+	}
+
+	defer res.Close()
+
+	buffer, err := io.ReadAll(res)
+	if err != nil {
+		f.logger.Errorf("ReadAll Failed. Unable to read json file: %v", err)
+		return nil, err
+	}
+
+	reader := bytes.NewReader(buffer)
+
+	decoder := json.NewDecoder(reader)
+
+	// Peek the first JSON token to determine the type
+	token, err := decoder.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	if d, ok := token.(json.Delim); ok && d == '[' {
+		return &jsonReader{decoder: decoder, token: token}, err
+	}
+
+	// Reading JSON object
+	decoder = json.NewDecoder(reader)
+
+	return &jsonReader{decoder: decoder}, nil
```

Comment:
err is nil when checking the token

Please consider doing this


```suggestion
	token, err := decoder.Token()
	if err != nil {
		return nil, err
	}

	jsonReader := jsonReader{decoder: decoder}
	if d, ok := token.(json.Delim); ok && d == '[' {
		jsonReader.token = token
	}

	return jsonReader, nil
```

Reply:
I have to again decode the bytes in the reader into the decoder... because while checking the token the offset moves from 0 to 1.  Then for a normal json object it results in error while parsing.

Reply:
Oh, OK. It's indeed something unexpected.

Then maybe adding a comment about the fact the new decoder need to be created because of that